src/fdb5/tools/fdb-hammer.cc

```diff
@@ -111,6 +111,339 @@ Key shortFDBKey(const Q& step, const R& member, const S& level, const T& param)
+
+            /// the leader PID is written into the file
+            SYSCALL(::close(fd));
+            std::unique_ptr<eckit::DataHandle> fh(pid_file_.fileHandle());
+            eckit::HandleStream hs{*fh};
+            fh->openForWrite(eckit::Length(sizeof(long)));
+            {
+                eckit::AutoClose closer(*fh);
+                hs << (long)::getpid();
+            }
+        }
+        else {
+
+            /// otherwise, the process is a client
+
+            /// the leader PID is read from the file
+            std::unique_ptr<eckit::DataHandle> fh(pid_file_.fileHandle());
+            eckit::HandleStream hs{*fh};
+            long pid;
+            eckit::Length size = fh->openForRead();
+            {
+                eckit::AutoClose closer(*fh);
+                /// the PID file may be empty if trying too soon
+                if (size == eckit::Length(0)) {
+                    fh->close();
+                    ::sleep(1);
+                    size = fh->openForRead();
+                }
+                if (size == eckit::Length(0)) {
+                    throw eckit::SeriousBug("Found empty PID file. Leader too slow or manual remove required.");
+                }
+                hs >> pid;
+            }
+            pid_t leader_pid = (long)pid;
+
+            /// the leadr PID is checked to exist. If not, clean up and
+            /// restart election procedure
+            if (::kill(leader_pid, 0) != 0) {
+                try {
+                    pid_file_.unlink();
+                }
+                catch (eckit::FailedSystemCall& e) {
+                    if (errno != ENOENT)
+                        throw;
+                }
+                continue;
+            }
+        }
+
+        leader_found = true;
+    }
+}
+
+void Barrier::init() {
+
+    electLeaderProcess();
+
+    // if this process is the intranode leader process and there are multiple nodes,
+    // establish TCP connections between all intranode leader processes.
+
+    if (!is_leader_process_ || nodes_.size() == 1) {
+        init_ = true;
+        return;
+    }
+
+    if (hostname_ == nodes_[0]) {
+
+        // this is the leader node
+        is_leader_node_ = true;
+
+        // accept as many connections as peer nodes, on the designated port.
+
+        eckit::net::TCPServer server(port_);
+
+        for (int i = 0; i < server_connections_.size(); ++i) {
+            server_connections_[i] = server.accept("Waiting for connection", max_wait_);
+        }
+
+        // the TCPServer is auto-closed
+    }
+    else {
+
+        /// @note if the server is not yet listening, the client connection wil fail and will be retried every second
+        /// until timeout. This could be adjusted to retry more frequently.
+
+        eckit::net::TCPClient client;
+        client_connection_ = client.connect(nodes_[0], port_, max_wait_, 0, 1);
+    }
+
+    init_ = true;
+}
+
+void Barrier::fini() {
```

Comment:
fini() is not a sensible (English) name for the function.

I don't see why these methods shouldn't be in a destructor, to make the object inherently safe if the surrounding code calsl exceptions.